### PR TITLE
cli: implement spec command

### LINKED ISSUE #188: Implement the spec command
The runtime does not implement the `runc spec` command that generates a basic `config.json`.

### PR DESCRIPTION
**Pull request by caoruidong**
Add spec command that  that generates a basic config.json for kata.

fixes #188

Signed-off-by: Ruidong Cao <caoruidong@huawei.com>

### CODE REVIEW

**Comment by caoruidong** [comment]
Sorry, forget to update dep. Wait

File: cli/spec.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
+package main
```
Comment (devimc) [inline cli/spec.go]: missing license

File: cli/spec.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+			}
+		}
+		if err := checkNoFile(specConfig); err != nil {
+			return err
+		}
+		data, err := json.MarshalIndent(spec, "", "\t")
+		if err != nil {
+			return err
+		}
+		if err := ioutil.WriteFile(specConfig, data, 0666); err != nil {
```
Comment (devimc) [inline cli/spec.go]: pure evil  ehh :smile:

File: cli/spec.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+			Name:  "bundle, b",
+			Value: "",
+			Usage: "path to the root of the bundle directory",
+		},
+		cli.BoolFlag{
+			Name:  "rootless",
+			Usage: "generate a configuration for a rootless container",
+		},
+	},
+	Action: func(context *cli.Context) error {
```
Comment (devimc) [inline cli/spec.go]: please add an unit test for this

File: cli/spec.go (hunk 2 above)
Comment (caoruidong) [reply cli/spec.go]: Change mode to 640?

File: cli/spec.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+	Flags: []cli.Flag{
+		cli.StringFlag{
+			Name:  "bundle, b",
+			Value: "",
+			Usage: "path to the root of the bundle directory",
+		},
+		cli.BoolFlag{
+			Name:  "rootless",
+			Usage: "generate a configuration for a rootless container",
+		},
```
Comment (WeiZhang555) [inline cli/spec.go]: I'm not sure if we can support "rootless" container in kata in future, as I know this is a runc feature that allows you to run a container without root privilege.

This flag needs to be removed if it's not supported yet.

File: cli/spec.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+	},
+	Action: func(context *cli.Context) error {
+		spec := specconv.Example()
+
+		rootless := context.Bool("rootless")
+		if rootless {
+			specconv.ToRootless(spec)
+		}
```
Comment (WeiZhang555) [inline cli/spec.go]: This part needs to be removed if `rootless` container isn't supported yet

**Comment by WeiZhang555** [comment]
I'm wondering if these vendor files added by `dep` tool? Please update the vendor with `dep` tool, one reference link is: 
https://github.com/kata-containers/runtime/pull/141

**Comment by bergwolf** [comment]
Could you add a test case the makes sure we can run a kata container using the default spec? There is [an example in runv] (https://github.com/hyperhq/runv/blob/master/tests/go-integration/test_data/Makefile) for creating a busybox bundle that might help you.

**Comment by caoruidong** [comment]
@bergwolf Sure. Thanks. I'm working on this.
@WeiZhang555 Yes.  I will update with dep together with unit test.

File: cli/spec.go (hunk 2 above)
Comment (devimc) [reply cli/spec.go]: runc uses 0664

Comment (WeiZhang555) [reply cli/spec.go]: I think runc is using 0666: https://github.com/opencontainers/runc/blob/master/spec.go#L110

But I don't know why, I think 0640 could be enough.

Comment (devimc) [reply cli/spec.go]: that's weird

```
$ docker-runc spec
$ ls -l config.json
-rw-rw-r-- 1 X X 597 Apr 11 10:33 config.json
```

Comment (sboeuf) [reply cli/spec.go]: Yes, let's do better than `runc` here, `0640` will be enough.

File: cli/spec.go (hunk 1 above)
Comment (caoruidong) [reply cli/spec.go]: Done

File: cli/spec.go (hunk 2 above)
Comment (caoruidong) [reply cli/spec.go]: Done

File: cli/spec.go (hunk 4 above)
Comment (caoruidong) [reply cli/spec.go]: Done

**Comment by caoruidong** [comment]
@WeiZhang555 Vendor updated.
@bergwolf After testing, the default spec can work with busybox. Do you mean that add a test case in spec_test.go to explicitly call kata-runtime run with default spec and busybox bundle ?

**Comment by bergwolf** [comment]
@caoruidong I don't think we can do it in `spec_test.go`. I was thinking some place in https://github.com/kata-containers/tests. It does not block this PR though.

**Comment by caoruidong** [comment]
@bergwolf So I think test_data/Makefile should be moved to https://github.com/kata-containers/tests.

**Comment by bergwolf** [comment]
@caoruidong Feel free to use it when adding new tests. I just pointed to the runv tests for your reference.

**Comment by sboeuf** [comment]
@devimc PTAL.

**Comment by caoruidong** [comment]
@jodh-intel @WeiZhang555 PTAL

**Comment by WeiZhang555** [comment]
LGTM after rebase.
Nice patch!

**Comment by caoruidong** [comment]
Rebased!

**Comment by caoruidong** [comment]
Can this be merged?

**Comment by jodh-intel** [comment]
Done on https://github.com/kata-containers/documentation/pull/48/commits/2f2956e38c6668e70973fbc9b694e6651ae4a70c.

**Comment by jodh-intel** [comment]
Tests fine in stand-alone mode fwiw (using the strategy below for `kata-runtime`):

- https://github.com/clearcontainers/runtime/wiki/Running-standalone

**Comment by grahamwhaley** [comment]
The only thing stopping the merge is the red codecov reports - if we are happy with that @jodh-intel , then let's merge...

File: cli/spec.go [hunk 6]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) 2014,2015,2016,2017 Docker, Inc.
+// Copyright (c) 2018 Huawei Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (grahamwhaley) [inline cli/spec.go]: Ah, I thought the standard for this project (the whole of kata) was the SPDX style copyright like:
https://github.com/kata-containers/runtime/blob/master/cli/fatal.go#L3
but, I note we have some non-SPDX copyright in the code base already, such as:
https://github.com/kata-containers/runtime/blob/master/cli/config.go#L3

@egernst @sboeuf @jodh-intel , thoughts?
And, maybe we can add a static check to catch any future deviations (whichever way we go) to the CI - maybe.

Other than that, this copyright looks correct to me afaict - the original code coming from https://github.com/opencontainers/runc/blob/master/spec.go I believe. That original file does not have an explicit copyright (it relies on the project wide LICENSE file I believe), so I presume we looked at the git logs to get the date range in the above - which I believe is fine and correct.

Maybe we can add a URL reference to that original file just below the copyright to make the historical origins clear?

Comment (grahamwhaley) [reply cli/spec.go]: btw, if we do all agree on SPDX (which we did I think at project inception, but we probably did not document well enough, and have not always caught transgressions), then I'm happy to Issue/PR to update/fix them all etc.

Comment (jodh-intel) [reply cli/spec.go]: Good spot @grahamwhaley! Yep, I agree - we should ensure all our files have the SPDX header where possible.

I've raised https://github.com/kata-containers/tests/issues/240 to add a CI check to enforce this for new / modified files, but would be great to have the existing codebase "SPDX-ified" :smile:

**Comment by caoruidong** [comment]
@grahamwhaley It's green last time... I'll update aggin. 
@jodh-intel SPDX-ing together.

**Comment by jodh-intel** [comment]
yep - let's do this! ;)

**Pull request merged**